### ISSUE TITLE
[OSX] Client crashes when spectating games

### ISSUE THREAD

**Issue by dayvough** [ValveSoftware/Dota-2#922]
After the June 25, 2016 patch, I can no longer spectate the TI6 qualifier matches. I click the "Watch In-game" and it crashes.

Mac OS X 10.11.4
Macbook Pro Retina 15" 2012

**Comment by Turmfalke2** [comment]
I am having similiar issues on linux, but it is not for all games.  
From dmesg:
[Jun25 10:30] GLRenderThread[26552]: segfault at 555f4449e100 ip 0000555f4449e100 sp 00007f05166ec708 error 15
[Jun25 10:31] traps: GLRenderThread[27890] general protection ip:7f9d59dee0f6 sp:7f9d5f517710 error:0 in libscaleformui_4_gl.so[7f9d59811000+a09000]
[Jun25 10:32] traps: GLRenderThread[27980] general protection ip:7f336c4c20f6 sp:7f3371c4a710 error:0 in libscaleformui_4_gl.so[7f336bee5000+a09000]

E: http://pastebin.com/BheKcmkS

**Comment by hrzhu** [comment]
Possibly the same problem as #918, but I have this problem even before June 25 patch.

**Comment by dayvough** [comment]
@hrzhu: I have been able to watch the Open Qualifiers in client with problem though.

**Comment by Turmfalke2** [comment]
To me it feels like it is related to drafting phase, it feels like I can reliably spectate games after the 00:00 minute mark. Unfortunately that is a bit of an issue with the compendium predictions.
E:Can't watch Alliance vs Team Secret, but Virtus.pro vs Team Empire is fine.
E2: Was able to join Alliance vs Team Secret at the end of the drafting phase.

**Comment by ramma** [comment]
Same here on Linux. It seems to happen when the draft is in progress. I only manage to load near the end of the drafts and any point after. When the draft screen does load, undrafted heroes have purple/pink boxes where spell images would be, and a pile of strength icons on the right of where the portrait would be.

**Comment by dayvough** [comment]
Update: I loaded into a game with a draft on and I was able to watch Secret vs Alliance. My previous attempts were when I loaded a game that was already past the bans.

Can also confirm that I'm also getting the weird strength icons + purple/pink boxes @ramma is talking about.

**Comment by gu3st** [comment]
If picks/bans have ended (players have loaded into the match), I very very seldomly crash.

If I try to join a game that's mid-draft, I have a extremely high crash rate. I have a bunch of crash dumps from these as well.

These dumps are from Linux, but the same thing happens in OS X, just with less frequency.

I'm running Ubuntu 16.04, but I had the same issue in 14.04.  Unity as my Desktop Environment.  

[crash_20160625194822_1.dmp.txt](https://github.com/ValveSoftware/Dota-2/files/333526/crash_20160625194822_1.dmp.txt)
[crash_20160626002642_1.dmp.txt](https://github.com/ValveSoftware/Dota-2/files/333527/crash_20160626002642_1.dmp.txt)
[crash_20160626003505_1.dmp.txt](https://github.com/ValveSoftware/Dota-2/files/333528/crash_20160626003505_1.dmp.txt)
[crash_20160626010423_1.dmp.txt](https://github.com/ValveSoftware/Dota-2/files/333529/crash_20160626010423_1.dmp.txt)

**Comment by Turmfalke2** [comment]
The strength icons and purple boxes are unrelated to this bug, they have been like that ever since we got reborn(?) and I am pretty sure they have been reported here before.

**Comment by rudmint** [comment]
Same issue on Linux

Jun 26 15:19:24 razorwolfelite kernel: [1655746.367483] GLRenderThread[25589]: s
egfault at 10 ip 00007f325e6c40f6 sp 00007f3263ddf710 error 4 in libscaleformui_
4_gl.so[7f325e0e7000+a09000]

Jun 26 15:20:26 razorwolfelite kernel: [1655808.275466] traps: GLRenderThread[25
963] general protection ip:7fe8e3a8a0f6 sp:7fe8e9166710 error:0 in libscaleformu
i_4_gl.so[7fe8e34ad000+a09000]

Jun 26 15:24:57 razorwolfelite kernel: [1656079.286565] traps: GLRenderThread[26
092] general protection ip:7f50688ab0f6 sp:7f506df87710 error:0 in libscaleformu
i_4_gl.so[7f50682ce000+a09000]

Sys info:

Processor Information:
    CPU Vendor:  GenuineIntel
    CPU Family:  0x6
    CPU Model:  0x3c
    CPU Stepping:  0x3
    CPU Type:  0x0
    Speed:  3400 Mhz
    4 logical processors
    2 physical processors
    HyperThreading:  Supported
    FCMOV:  Supported
    SSE2:  Supported
    SSE3:  Supported
    SSSE3:  Supported
    SSE4a:  Unsupported
    SSE41:  Supported
    SSE42:  Supported
    AES:  Supported
    AVX:  Supported
    CMPXCHG16B:  Supported
    LAHF/SAHF:  Supported
    PrefetchW:  Unsupported

Network Information:
    Network Speed:  

Operating System Version:
    Ubuntu 16.04 LTS (64 bit)
    Kernel Name:  Linux
    Kernel Version:  4.4.0-22-generic
    X Server Vendor:  The X.Org Foundation
    X Server Release:  11803000
    X Window Manager:  Compiz
    Steam Runtime Version:  steam-runtime-release_2016-02-23

Video Card:
    Driver:  NVIDIA Corporation GeForce GTX 750 Ti/PCIe/SSE2

```
Driver Version:  4.5.0 NVIDIA 361.42
OpenGL Version: 4.5
Desktop Color Depth: 24 bits per pixel
Monitor Refresh Rate: 60 Hz
VendorID:  0x10de
DeviceID:  0x1380
Number of Monitors:  1
Number of Logical Video Cards:  1
Primary Display Resolution:  1920 x 1080
Desktop Resolution: 1920 x 1080
Primary Display Size: 20.08" x 11.30"  (23.03" diag)
                                        51.0cm x 28.7cm  (58.5cm diag)
Primary Bus: PCI Express 16x
Primary VRAM: 2048 MB
Supported MSAA Modes:  2x 4x 8x 16x 
```

Sound card:
    Audio device: Realtek ALC887-VD

Memory:
    RAM:  16006 Mb

**Comment by gdrewb-valve** [comment]
If you go into the Steam client and validate your install does it find any issues?

**Comment by gu3st** [comment]
No issues found when validating.

**Comment by Yabk** [comment]
Same problem, tried validating, it didn't find any issues.

**Comment by Sebby91** [comment]
My game always freeze for a few seconds between when the heroes are chosen (picks/bans) and players are choosing the hero which they play. Most of the time after the freeze, it'll just go into game but sometimes (especially after watching several games in a row) it will crash.

**Comment by gu3st** [comment]
The freezing => crash/disconnect is something that happens often to people with HDDs at the start of CM games.  I play in a league, and nearly every game starts with a few disconnects as a result. I think (in the disconnecting case) there's something that is blocking the network connection which causes it to timeout, as it didn't happen after I moved my stuff to SSD.

**Comment by Yabk** [comment]
Nope, I'm on SSD. The problems started when i upgraded to Ubuntu 16.04 (from 14.04)

**Comment by gdrewb-valve** [comment]
If you look at the steam-for-linux section you'll find a variety of problems when running Steam on 16.04.  Are you using any of those workarounds?

**Comment by gu3st** [comment]
I'm not using any workarounds. Taking a cursory glance, most of the workarounds seem to be AMD GPU specific, and I'm on nVidia.  The crashing that I'm experiencing when loading into a match in the draft happened in 14.04 as well, and still occasionally happens in OS X as well.

**Comment by gdrewb-valve** [comment]
There's a problem that occurs on all platforms that the larger team is working on, but there may be multiple problems involved.

**Comment by Yabk** [comment]
I just managed to connect to a game that was in pregame. Seems like I can't connect to games that are already in picking phase. If I connect during pregame seems like everything is working fine during the picking phase.

**Comment by gdrewb-valve** [comment]
There's definitely a problem here and Valve is getting a few reports of this kind of crash per day but we still haven't found what is going wrong.  This will likely take quite some time to track down.

**Comment by gdrewb-valve** [comment]
For people using the stable Steam client and willing to try the beta Steam client does that change anything?  Does disabling the Steam overlay for Dota change anything?

**Comment by ramma** [comment]
@gdrewb-valve I've been on the Steam Beta branch for a while. Since this is so time dependent I haven't been able to swap clients fast enough to test both clients with the same match.

Disabling the overlay did avoid the crash for me though.

I also noticed that only captains mode games are crashing. Whether the game is part of a tournament or not doesn't seem crucial, but I was only able to test with 1 non-tournament captains mode game.

**Comment by gdrewb-valve** [comment]
Interesting, thanks for the extra info.  We'll have to see if other people have similar results and that might narrow down the possible culprits.

**Comment by Turmfalke2** [comment]
It is Ti (or almost) and the bug is still around:
[1484306.232123] traps: GLRenderThread[28531] general protection ip:7fd7f8b2b106 sp:7fd7fe2b2750 error:0 in libscaleformui_4_gl.so[7fd7f854e000+a09000]
[1484348.981886] traps: GLRenderThread[29059] general protection ip:7f11a0189106 sp:7f11a58b1750 error:0 in libscaleformui_4_gl.so[7f119fbac000+a09000]

**Comment by gdrewb-valve** [comment]
@Turmfalke2, does turning off the Steam overlay help?

**Comment by gdrewb-valve** [comment]
Also do you have a match ID of a specific match that crashed for you?

**Comment by Turmfalke2** [comment]
@gdrewb-valve  where do I turn off the steam overlay? Also I can reproduce it with pretty much any captains mode game, as mentioned earlier it seems to depend on the state of the draft.
I think it was one of the Ehome games, let me check the time stamp. Ok according to dmesg it was 22:53:25 CEST, which is 20:53:25 UTC.
The uptime of my machine here is 17 days, but I don't suspend it. Therefore the time stamp should be pretty accurate. 

If you want be to reproduce it in a debugger, I am pretty confident that I could reproduce it given a captain's mode game.

**Comment by gdrewb-valve** [comment]
In the Steam client if you right-click on Dota in your app list you can get the properties.  In the General tab there's a checkbox for the overlay.

I was hoping to get a match ID so that I could check with exactly the same match so if you have one that'd be great.  Otherwise I'll pick a captain's mode game at random.

**Comment by Turmfalke2** [comment]
I will give it a try.

 I don't think it depends on the match ID and I am pretty sure it is not something you can reproduce from a replay. It depends more on the state of the draft.

E: Is there a way for me to get a random captain's mode game? Seems like OG vs LGD is supposed to start soon, I will try to reproduce it with that one.

**Comment by gdrewb-valve** [comment]
Do you mean an automatic random selection of a captain's mode game?  I don't think there is one, I was just going to check around manually.

**Comment by Turmfalke2** [comment]
Was hoping there was a hidden feature to filter the games on the watch/live tab. Appears like there are some delays with the TI games?

E: Checking the "Friends Only" checkbox when no friend is playing hides it and appears to constantly loop. Seems like restarting the client is required to recover from that.

**Comment by Turmfalke2** [comment]
Ok, success I crashed.
[Mi Aug  3 18:35:51 2016] GLRenderThread[6494]: segfault at 10 ip 00007f68c6a89106 sp 00007f68cc210750 error 4 in libscaleformui_4_gl.so[7f68c64ac000+a09000]
[Mi Aug  3 18:37:07 2016] GLRenderThread[6622]: segfault at 10 ip 00007f0c7930a106 sp 00007f0c7ea32750 error 4 in libscaleformui_4_gl.so[7f0c78d2d000+a09000]
They look a bit different from the ones yesterday, but the first one was with overlay activated and the second without it.

E: uh.. this one is new.
[Mi Aug  3 18:42:07 2016] GLRenderThread[6940]: segfault at 0 ip           (null) sp 00007f5ecfae8748 error 14

**Comment by gdrewb-valve** [comment]
OK, thanks for confirming it happens even without overlay.  We'll have to try and catch it here at Valve to make more progress.  Given that it may only occur at certain points during the draft of live games it may take a while to isolate.

**Comment by rudmint** [comment]
I'm currently crashing attempting to watch the NAVI group stage match.  One thing I notice which may or may not be of consequence is that the load screen is always the same before a crash, even though I have 100+ load screens set on random.

**Comment by Turmfalke2** [comment]
Just in case this really matters - I am not using a custom load screen.

**Comment by rudmint** [comment]
I'm crashing right now, attempting to watch the NAVI group stage match.

**Comment by gdrewb-valve** [comment]
If you look at the Steam client system information (under the Help menu) does it have failure reports that correspond to the times you're crashing?  If they have crash ID's post them here and I can see if they're the same thing or different.

**Comment by rudmint** [comment]
I do have them, but they appear to be discarded;

```
Wed Aug  3 17:38:01 2016 GMT: file ''/tmp/dumps/crash_20160803133727_1.dmp'', upload yes: ''Discarded=1''
Wed Aug  3 17:40:12 2016 GMT: file ''/tmp/dumps/crash_20160803133939_1.dmp'', upload yes: ''Discarded=1''
```

**Comment by gdrewb-valve** [comment]
Most will be discarded, unfortunately.  If you do end up with crash IDs post them here.  So far I haven't been able to watch at exactly the right moment to catch the problem.  Additionally catching a single instance is probably not going to be sufficient to isolate the bug so this one is going to be very difficult.

As mentioned elsewhere I also have a suspicion that this is related to a particular crash that we get more of during tournaments.  That's nicely correlated with an increase in spectation of captain's mode games.  That crash happens on all platforms and the larger team has been looking at for a while with no succcess.  If I get a crash ID I can see if it is really the same thing.

**Comment by gdrewb-valve** [comment]
I've tried grabbing some of the TI replays available and, as suspected, they haven't shown a problem yet.

**Comment by Turmfalke2** [comment]
@gdrewb-valve: I have one that didn't get discarded:
Wed Aug  3 16:41:00 2016 GMT: file ''/tmp/dumps/crash_20160803183935_1.dmp'', upload yes: ''CrashID=bp-72b045a3-da27-4672-a2fc-420da2160803''
Can you find the dump with this id on a servers of yours or shall I send you it?
I assume it is the dump related to the "segfault at 0 ip (null)" earlier.

**Comment by gdrewb-valve** [comment]
Thanks, that's the all-platforms crash I believe is the underlying culprit.

**Comment by gdrewb-valve** [comment]
As you've seen from your own reports the crash is in Scaleform, which is a third-party library from Autodesk and isn't Valve's code.  The problem may be in some interaction with Valve's code but it makes the whole thing even more difficult since the actual piece of code that's dying is not something people at Valve are familiar with.  The larger team is going to keep grinding away but I wouldn't expect any quick resolution here (unless we're lucky).

**Comment by Turmfalke2** [comment]
Yes.. I can see some issues with scaleform, they aren't even using the correct ssl certificate for websites. Really wouldn't want to be the guy who decided to use scaleform.
Which version of scaleform is dota using?

**Comment by gdrewb-valve** [comment]
4.2

**Comment by gdrewb-valve** [comment]
The Aug. 5 update has an experiment to try and gather some information.  If you add '+r_sync_scaleform_frame_and_render 1' to your launch options does it make any difference?  This is a long shot and most likely will not help, just to set expectations.

**Comment by Turmfalke2** [comment]
I guess I can try that on Monday, unless you know of any earlier captain's mode games.

**Comment by ramma** [comment]
@Turmfalke2 I've just been keeping trackdota.com up waiting for tournament matches to end. Much easier than scrolling through live matches with no filters, but far from ideal. This would be much easier to test with a filter or API function to list the newest pub matches of a particular mode.

**Comment by ramma** [comment]
@gdrewb-valve Finally found a captains mode game to test with (match id 2552333508). The launch option did not help.

This could be unrelated, but after crashing with that launch option I can't spectate that game again. I get the message "Watch game status Could not watch this game at this time", with or without the launch option and persisting through restarts. Other matches work fine.

**Comment by gdrewb-valve** [comment]
OK, thanks for trying it out.  We'll keep looking for a fix.

**Comment by rudmint** [comment]
Surprisingly haven't had this issue since I last posted (7 days).  Currently attempted to connect to Fnatic Alliance match and have a couple crash reports if you still want them.

```
Wed Aug 10 17:49:01 2016 GMT: file ''/tmp/dumps/crash_20160810134900_57.dmp'', upload yes: ''CrashID=bp-4642b628-ab87-42e7-9d0f-26a832160810''

Wed Aug 10 17:49:33 2016 GMT: file ''/tmp/dumps/crash_20160810134906_1.dmp'', upload yes: ''CrashID=bp-3a074dee-afd3-4410-9ea9-d3bc52160810'' 
```

**Comment by gdrewb-valve** [comment]
One of your issues should be fixed in the current Steam client beta, are you running that or the stable client?

**Comment by rudmint** [comment]
I am currently using the stable client, I'll download the beta and see what happens.

**Comment by rudmint** [comment]
I downloaded the beta client, trying to connect to MVP.P Wings game now,

 Wed Aug 10 23:40:14 2016 GMT: file ''/tmp/dumps/crash_20160810165953_1.dmp'', upload yes: ''CrashID=bp-3624adf2-11f8-4cdf-9f2f-014be2160810''

As soon as the draft ended, I was able to connect.

**Comment by gdrewb-valve** [comment]
That's the Scaleform crash, so hopefully you're no longer seeing the Steam client crash,

**Comment by jazztickets** [comment]
Using vulkan seems to avoid the crash. Might be a good workaround for linux users for now.

Specs:
https://gist.githubusercontent.com/jazztickets/ec290c38deb7fa8db404/raw/bbe72e82bc6c339ed0322dc0d26cbfdbf3daf69a/gistfile1.txt

**Comment by admshao** [comment]
This just happened today again when broadcasting a Professional Tier tournament and i had to capture another brodcaster video so i couldn't complete miss the draft.. It's a pretty serious issue guys. In 2 days a Premium tournament is about to start and in a couple of weeks a Major!

Do you recommend me to drop my linux setup while this is being worked on? Let me know if any other info is needed.

Just in case all those infos from the closed issue might go unnoticed. 

"It's been months that this error keeps appearing. Some days more frequently than others.
Currently under ubuntu 16.10, Gnome 3.22, NVidia 1080 with 370.28 drivers.
I don't have much information to give other than that it happens when spectating a tournament match while game is already in the draft and on Captains Mode. I wouldn't discard it happening while game is loading before the draft screen tho.

It is 100% reproductible ONCE it happens for the first time for that given game. Cpu load is far from loaded, always sitting around 30%. Dota video settings is all on max with v-sync off.

It doesn't matter any cosmetic about terrain or weather.

Doesn't matter if Vulkan or GL. Neither the Graphics driver version (tested under 368, 370, 375).
Also tried a couple o kernel versions and happens on all. (4.4 - 4.8).

I always experience this while co-casting from DotaTV. When on a caster slot inside lobby it is totally fine. So strictly to DotaTV, be it a ticket game or not.

[18501.081317] VKRenderThread[1145]: segfault at 88 ip 00007f34fe8dae5d sp 00007f35050398f0 error 4 in libscaleformui_4_vulkan.so[7f34fe26d000+b04000]
[18541.059219] traps: GLRenderThread[17087] general protection ip:7fa0429d1bd6 sp:7fa048019750 error:0 in libscaleformui_4_gl.so[7fa0423f9000+a03000]
[18588.599505] VKRenderThread[18912]: segfault at 88 ip 00007f2799b39e5d sp 00007f27a02988f0 error 4 in libscaleformui_4_vulkan.so[7f27994cc000+b04000]

System info with CrashID from today -> https://gist.github.com/admshao/b3c1b60add3e2b4b29bedb77b875a7ac

Console.log from past couple of months or so including crashs from today and bunch of other days ->
https://gist.github.com/admshao/d1a09887521d07432f32a4b119b10c49"

**Comment by gdrewb-valve** [comment]
As mentioned in the comments here at least one of the crashes involved happens on Windows also and is being checked out by the larger team.  It's also in non-Valve code so finding and fixing the issue is going to be extremely difficult.  We're not aware of any workarounds that work for everybody in all cases, but some people have had luck using Vulkan.

**Comment by FlaminSarge** [comment]
I've too have experienced this on OSX when attempting to spectate a game that is in its draft phase. If I do manage to get through to spectate the draft, the game's performance (and in some cases, my Mac's performance) tanks during the draft phase (but once the actual gameplay starts it returns to normal).

Where would the relevant crash dumps and logs be for this problem on OSX? I'm having trouble finding them.

In addition, does `con_logfile` function on Dota 2?

**Comment by kaiserprogrammer** [comment]
I have the same issue when  "-map dota" is set in launch options, ~~removing it fixed this issue for me.~~
It worked only once. So my bad.
~~Somehow messing with the launch options "fixed" it for me.~~

**Comment by gdrewb-valve** [comment]
7.00 has reduced the usage of Scaleform and may affect this.

**Comment by admshao** [comment]
Nothing has changed for me. With -vulkan i'm unable to watch a game in the draft phase of Captains Mode. Like 9 out of 10 times. Always crashing on vulkan scaleform lib in dmesg. Same issue as before 7.00

**Comment by gdrewb-valve** [comment]
Unfortunately Captain's Mode is still using scaleform so the crash is still going to happen.  Eventually the larger team will convert it as they've done with other game modes.

**Comment by fleetcommand** [comment]
Oh my issue was really a duplicate of this. So posting the crash dump if it helps anything. But I doubt it since - as it was mentioned above - some non-valve code is also affected in this manner.

Anyway.. it can't hurt, so here it is:
[dota2_2017_0131_194955_0_accessviolation.zip](https://github.com/ValveSoftware/Dota-2/files/742946/dota2_2017_0131_194955_0_accessviolation.zip)

**Comment by kisak-valve** [comment]
Does anyone still encounter this issue with the updated captain's mode UI in the [2017-03-15 Dota 2 update](http://www.dota2.com/bladeformlegacy)?

**Comment by gdrewb-valve** [comment]
I believe the CM spectating UI didn't change, just the live UI.

**Comment by kisak-valve** [comment]
Thanks for the factoid @gdrewb-valve.

**Comment by admshao** [comment]
Hey guys. it still crashes for me like every day i have to cast the games. 1 or 2 months during kiev major and Manilla masters it started to crash on Windows too for my co-workers. TI 7 is just around and we'll cover the event. We hope it will be fixed by then since UI changes are coming to Ability Draft with this hunting us for months on Captains Mode.

As soon as the scaleform UI goes out from the draft we can join the game with np

**Comment by gdrewb-valve** [comment]
This is a Scaleform issue and not a Linux/Mac issue so I'm not sure what's happening in the larger team.   I believe the long term plan is to switch away from Scaleform but that'll happen when it happens, nobody knows when exactly that'll occur.

**Comment by jazztickets** [comment]
I'm assuming the new spectator draft screen for TI doesn't use scaleform at all? It no longer crashes for me on linux.

**Comment by gdrewb-valve** [comment]
Correct, more usage of Scaleform has been removed.  There are still some situations where it can be used but things are closer to running entirely without it.

**Comment by jazztickets** [comment]
Well, I dare say this issue can finally be closed unless more reports show up.

**Comment by kovetskiy** [comment]
Today I have spectated match and dota instantly crashed, message in dmesg:
```
[  399.335768] GLRenderThread[5441]: segfault at 0 ip 00007f28095ba5d6 sp 00007f280dbc1ba0 error 4 in libscaleformui_4_gl.so[7f2809443000+9e7000]
```

**Comment by gdrewb-valve** [comment]
Scaleform has been completely removed so any spectating issues from here on out are new issues.
